Refactor IPv6 header length handling

ozhuraki opened the pull request:
Change the length to uint16_t and work with it through standard htons/ntohs() macros.

jukkar commented on tests/net/6lo/src/main.c:
```diff
@@ -397,7 +396,6 @@ static struct net_6lo_data test_data_1 = {
 	.ipv6.vtc = 0x60,
 	.ipv6.tcflow = 0x00,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
```
Even if it makes not difference here (as the variable is static), for consistency please set .len = 0 here like you did for couple of these testing variables below.

ozhuraki replied:
OK, corrected.

jukkar reviewed (changes requested):
LGTM, one minor tweak could be done.

pfalcon commented on subsys/net/ip/ipv6.c:
```diff
@@ -2123,11 +2118,12 @@ int net_ipv6_send_ns(struct net_if *iface,
 	net_icmpv6_set_ns_hdr(pkt, ns_hdr);
 
 	if (is_my_address) {
+		u16_t *len = &NET_IPV6_HDR(pkt)->len;
```
I'd say, this gets hairy and hardly an improvement. Please use standard var arithmetics, without pointers.

ozhuraki replied:
As for my taste, the original snippet is slightly overloaded visually and using pointer here makes it easier to parse what's going on. But, don't really want on to argue on such matters of taste, OK, corrected.

pfalcon commented on tests/net/6lo/src/main.c:
```diff
@@ -435,7 +432,7 @@ static struct net_6lo_data test_data_3 = {
 	.ipv6.tcflow = 0x21,
 	.ipv6.flow = 0x3412,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = 0,
```
Nitpick, but did you consider using htons(0) to avoid any concerns?

pfalcon commented on tests/net/6lo/src/main.c:
```diff
@@ -397,7 +396,6 @@ static struct net_6lo_data test_data_1 = {
 	.ipv6.vtc = 0x60,
 	.ipv6.tcflow = 0x00,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
```
Why here you remove this line completely, while bellow replace with ".ipv6.len = 0,".

pfalcon commented on tests/net/6lo/src/main.c:
```diff
@@ -582,7 +572,7 @@ static struct net_6lo_data test_data_11 = {
 	.ipv6.tcflow = 0x20,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = 0,
```
Same comments as above.

ozhuraki replied:
OK, corrected.

pfalcon reviewed (changes requested):
Thanks for going thru these, but I'd say some cleanup needed.

jukkar commented on tests/net/6lo/src/main.c:
```diff
@@ -397,7 +396,7 @@ static struct net_6lo_data test_data_1 = {
 	.ipv6.tcflow = 0x00,
 	.ipv6.flow = 0x00,
-	.ipv6.len = { 0x00, 0x00 },
+	.ipv6.len = htons(0),
```
IMHO this is totally useless, 0 is always 0 in every byte order. Well, we can go with this too as this makes no difference here.

finikorg reviewed (approved):
LGTM

jukkar commented on subsys/net/ip/ipv6.c:
```diff
@@ -2126,8 +2121,9 @@ int net_ipv6_send_ns(struct net_if *iface,
 				net_ipv6_unspecified_address());
-
-		NET_IPV6_HDR(pkt)->len[1] -= llao_len;
+		NET_IPV6_HDR(pkt)->len = ntohs(NET_IPV6_HDR(pkt)->len) -
```
> ntohs(NET_IPV6_HDR(pkt)->len) = htons(len);

What is this suppose to do, why is there ntohs() on left side?

pfalcon replied:
Fixed copy-paste error. It's supposed to do a typical read-modify-write pattern.

pfalcon replied:
> Sorry, my bad, just parsed it wrong. Corrected. Good catch.

Still not there, sorry. You now have:

~~~
NET_IPV6_HDR(pkt)->len = htons(NET_IPV6_HDR(pkt)->len -
						llao_len);
~~~

But you should take a ->len which is in networking order, convert to host order (you don't), subtract, convert back to netorder, store to ->. The best way to do it is apparently using an intermediate var, like my snippet in previous comment shows. Thanks.

ozhuraki replied:
My bad. Missed conversion in setup_headers(). Corrected. Take another look, please.

ozhuraki committed:
net: ipv6: Refactor IPv6 header length handling

Change the length to uint16_t and work with it
through standard htons/ntohs() macros.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

pfalcon reviewed (approved):
Thanks!
